crates/abd-clam/src/chaoda/graph.rs

```diff
@@ -0,0 +1,456 @@
+//! A `Graph` is a collection of `OddBall`s.
+
+use core::{cmp::Reverse, ops::Index};
+use std::collections::BinaryHeap;
+
+use distances::Number;
+use ndarray::prelude::*;
+use ordered_float::OrderedFloat;
+use rayon::prelude::*;
+
+use crate::{Dataset, Instance};
+
+use super::OddBall;
+
+/// A `Graph` is a collection of `OddBall`s.
+///
+/// Two `OddBall`s have an edge between them if they have any overlapping volume,
+/// i.e. if the distance between their centers is no greater than the sum of their
+/// radii.
+#[derive(Clone)]
+pub struct Graph<U: Number> {
+    /// The collection of `Component`s in the `Graph`.
+    components: Vec<Component<U>>,
```

Comment:
I broke the `Graph` into individual `Component`s because it significantly reduces the cognitive burden of the implementation of several CHAODA member methods.

Reply:
I really like this reorganization.